impl(bigquery): Logging decorator cleanup - Added DebugString member functions to …

jsrinnn opened the pull request:
…GetJobResponse and ListJobResponse so it can be used in the logging decorator.

Subsequent PR will contain the logging decorator changes that will use the logging from this PR.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131)
<!-- Reviewable:end -->

devbww reviewed (commented):
This seems to be digging the _"Use tracing options for request and response logging"_ hole of   #11122 even deeper.

Can we have `std::string DebugString(TracingOptions const&)` support for these non-proto request/response structs instead?

jsrinnn reviewed (commented):
Ack. I was planning on adding that in the decorator changes which is in the subsequent PRs. This change only has ostream logging.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRdsl1z283gREJ3hDZj:bjfcgzx)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
I am trying to be mindful of the PR sizes. 

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRdt5Fe9jY3r4Re7UYk:busc5tc)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

devbww reviewed (commented):
But I don't think we need `operator<<()` support for these any more than we need it for the protobuf request/responses.  What we do need is a way to add appropriately formatted versions of the request/responses to the debug log.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRe0MzPD3bZmfuDAaf1:bxfpqbo)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
I see. I would like some clarification here for my understanding

The ostream logging was added so in the decorator I can do something like this. 

 `GCP_LOG(DEBUG) << (ListJObsResponse Object or GetJobsResponse object)`

1) Are you saying for doing the above, ostream logging is not needed? I thought it was needed for the default log sink but maybe not?

2) Instead of operator<<, I need to add 

DebugString(ListJobsResponse, TracingOptions)
DebugString(GetJobResponse, TracingOptions)

and so on...?

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRe1cqRAu_BkFW3WQWw:bnn6wsq)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

devbww reviewed (commented):
> 1. Are you saying for doing the above, ostream logging is not needed?

Yes ... I can't see how it gets you towards the end goal.

> 2. Instead of operator<<, I need to add
> 
> DebugString(ListJobsResponse, TracingOptions)
> DebugString(GetJobResponse, TracingOptions)

Yes ... or as member functions if that is easier.

Then, in the decorator, you can `GCP_LOG(DEBUG) << DebugString(request, options)` like all the other services.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRe6UGjDXdKfOPfvoSZ:b-l62zo7)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

coryan commented:
I think the confusion is my fault, I suggested `operator<<()` because it also works for tests, and in GCS we do logging with it.  Both can be made to work, and I like having good errors in my tests.

devbww reviewed (commented):
Fair enough ... if we really need `operator<<()` on these for tests, etc., we can define it as `os << DebugString(payload, TracingOptions{})`.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NReA0N58iiOsTyTkv6k:b-telzsx)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
I am getting conflciting feedback. Please advise clearly what is the conclusion here and what needs to change in this PR.

1)  Should I modify this to DebugString() ?

2) Should I keep the operator<< ?

Thanks

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NReESy-5JNeR5dQ2pw-:b-tuqjkp)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
I am going to do the following changes tomorrow for this PR. I think it might achieve both yours and Carlos's suggestion.

1) I'll keep the operator<<  so it can be used in tests

2) I'll add the DebugString(<Request>/<response>, TracingOptions) member functions so we can use it in the logging decorator.

Thanks

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NReg4uj8f-t2cTzGMPH:ba5cl2z)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
Sorry, but quoting and repeating your previous comment is not helpful here as I was trying to get clarification around that. 

I'll go ahead and make the changes I have said before. I have mentioned them below (again) for reference

1) I'll keep the operator<< so it can be used in tests

2) I'll add the `DebugString(Request, TracingOptions)` and `DebugString(Response, TracingOptions)` member functions so we can use it in the logging decorator. There is already a `DebugString(string, TracingOptions)`

For future reference,

I  understand and appreciate Carlos and Darren's feedback because they make the effort to clearly explain what change is needed and the reason for the change.

With you, the feedback is either vague or conflicting with feedback that is already given to me by someone else. This has happened multiple times now. I have to go back and forth trying to to understand exactly what you mean. Given the amount of work involved in this project, this is not a good use of anyones time.

Please try to give actionable feedback.

Thank You

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRf-X54EQouElshF6Yf:bgwuds6)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

devbww reviewed (commented):
> Please try to give actionable feedback.

I always strive to do just that.  I apologize if it appears otherwise.  I'm happy to chat at your convenience to clear anything up.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRf92Cm9vGUCSN4VPSQ:buptz42)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
Do you agree with what I have listed as (1) and (2) in the previous comment regarding the changes, especially (2)? 

If so , then I can go ahead and implement that. If not then I can setup a meeting with you.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRfAtkB0584gdGc40oZ:bvfwd4a)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
Thank you Bradley. Thats all I was looking for. 

I'll go ahead and implement (2)  for DebugString(). I can remove (1)

For context,

From your earlier comments I got the impression I need to do (2) but the later comments seem to imply  (1) and I wasn't sure if it was an alternative or you were asking to implement it both ways. Hence the confusion.

Thanks.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NRhZvQvDNY6AXr6gjt0:bso77zg)* status: 0 of 6 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

jsrinnn commented:
fyi: I just rebased with main. Haven't implemented the latest code review feedback yet. Will send an updated comment once the feedback is implemented.

jsrinnn commented:
The latest commit for this PR contains the following changes:

1) Adds DebugString as member functiosn for the relevant response objects

2) Removed ostream methods

3) Modified unit tests according to the above

4) The logging decorator and request changes are not part of this PR. I'll follow them in a separate PR after this one.

Bradley, If you still feel the changes in the latest commit deviates from what you had in mind in a major way, then lets set up a meeting to discuss this. 

Lastly, please  understand that lot of API work is still to be done, so if any nits around formatting styles (related to logging) can be delayed to cleanup PRs at a future time, I would really appreciate that.

Thank You

jsrinnn reviewed (commented):
The latest PR commit does the following

Incorporated code review feedback from PR-11141 specifically

1) Added a custom delimeter that depends on tracing options so `single_line_mode` is heeeded.

2) Applied `truncate_string_field_longer_than` to individual string fields instead of whole string.

PTAL. 

Thanks!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NSCSZGF6p11dIXYCRRW:b9hgbhg)* status: 0 of 12 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

devbww commented on google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc:
```diff
@@ -42,6 +45,19 @@ StatusOr<BigQueryHttpResponse> BigQueryHttpResponse::BuildFromRestResponse(
   return response;
 }
 
+std::string BigQueryHttpResponse::DebugString(
```
We've added support for a `DebugFormatter` in "google/cloud/internal/debug_string.h" to help create consistently-formatted strings for request/response logging (consistent with the proto messages).

I updated the `DebugString()` functions for [`GetJobRequest`](https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/v2/minimal/internal/job_request.cc#L78-L86) and [`ListJobsRequest`](https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/v2/minimal/internal/job_request.cc#L88-L110) to use `DebugFormatter`.  Perhaps you could try it here.

You may have issues with repeated-ish fields (like `BigQueryHttpResponse.http_headers`), or sub-messages (like `GetJobResponse.http_response` and `ListJobsResponse.http_response`).  If so, it would be nice to understand them so that we can improve `DebugFormatter`.

devbww replied:
> You may have issues with ... sub-messages

I added `DebugFormatter` support for sub-messages, but it means that any `DebugString()` functions using this framework will now also take a leading "name" parameter, which will be used to produce the `name { ... }` part.

The intent is that the caller from the decorator will pass the full type name, while `SubMessage()` callers will pass the field name.

> You may have issues with repeated-ish fields

I'll look into that today.

None of this, of course, means you have to adopt `DebugFormatter` here.  We can address later.

devbww commented on google/cloud/bigquery/v2/minimal/internal/job_response.cc:
```diff
@@ -51,6 +54,17 @@ StatusOr<nlohmann::json> parse_json(std::string const& payload) {
   return json;
 }
 
+std::string DebugJobsString(std::vector<ListFormatJob> const& jobs,
```
It looks like this, and all the helper functions above, should be in an anonymous namespace.

jsrinnn reviewed (commented):
Addressed all feedback. Thank you, Bradley!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NSRggYe6_QvPJYumg0F:b-7jlxxz)* status: 0 of 12 files reviewed, 9 unresolved discussions (waiting on @devbww)

___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.h` line 22 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR6-r3-22:-NSNGco6AzPeEgvZN15A:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.h#L22)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This additional looks unnecessary.
</blockquote></details>

Done.
___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc` line 17 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR5-r3-17:-NSNGpKh4zetmlsZt-jJ:blfv0ym) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc#L17)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This additional looks unnecessary.
</blockquote></details>

It wasn't used previously but is now needed for the DebugFormatter
___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc` line 19 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR5-r3-19:-NSNGvgJ3wNzd2y90vcc:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc#L19)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This additional looks unnecessary.
</blockquote></details>

Done.
___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc` line 48 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR5-r3-48:-NSNRwNJ8fST9aNdmCUt:bseydwq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc#L48)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

We've added support for a `DebugFormatter` in "google/cloud/internal/debug_string.h" to help create consistently-formatted strings for request/response logging (consistent with the proto messages).

I updated the `DebugString()` functions for [`GetJobRequest`](https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/v2/minimal/internal/job_request.cc#L78-L86) and [`ListJobsRequest`](https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/v2/minimal/internal/job_request.cc#L88-L110) to use `DebugFormatter`.  Perhaps you could try it here.

You may have issues with repeated-ish fields (like `BigQueryHttpResponse.http_headers`), or sub-messages (like `GetJobResponse.http_response` and `ListJobsResponse.http_response`).  If so, it would be nice to understand them so that we can improve `DebugFormatter`.
</blockquote></details>

Done

Nice! Thank you, Bradley. This definetely makes the code formatting a lot better.

The changes in job_request files were very helpful for reference.  I have made similar formmater changes to all the response and related files. 

To me the nested output looks fine, but let me know if you think otherwise and maybe we can make them at a later point.
___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response_test.cc` line 103 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR7-r3-103:-NSNHEZl91AZ4BiiOpAx:b-73pfco) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response_test.cc#L103)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Unused?
</blockquote></details>

Done.

Yup. Thanks
___
*[`google/cloud/bigquery/v2/minimal/internal/job.cc` line 17 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRii6LcBN9IWdkQwfqK-r3-17:-NSNHQ3N5-OAiI3LPQWx:b-qyloti) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/job.cc#L17)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This include looks unnecessary.
</blockquote></details>

Done.

It wasn't used previously but is now needed for DebugFormatter changes.
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.h` line 23 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR9-r3-23:-NSNH_qW-JkTg_Er1wUu:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/job_response.h#L23)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This additional looks unnecessary.
</blockquote></details>

Done.
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 19 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR8-r3-19:-NSNHfCK-7HbGi0A7gzc:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L19)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This additional looks unnecessary.
</blockquote></details>

Done.
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 57 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR8-r3-57:-NSNHxZc0TpUYQXtt8K-:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L57)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like this, and all the helper functions above, should be in an anonymous namespace.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131)* status: 0 of 12 files reviewed, 9 unresolved discussions (waiting on @devbww)

___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc` line 48 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR5-r3-48:-NSRpvYP0P9ttQOJNYXs:b-yw1mi8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc#L48)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

> You may have issues with ... sub-messages

I added `DebugFormatter` support for sub-messages, but it means that any `DebugString()` functions using this framework will now also take a leading "name" parameter, which will be used to produce the `name { ... }` part.

The intent is that the caller from the decorator will pass the full type name, while `SubMessage()` callers will pass the field name.

> You may have issues with repeated-ish fields

I'll look into that today.

None of this, of course, means you have to adopt `DebugFormatter` here.  We can address later.
</blockquote></details>

Ack. 

Can you give approval for the changes in this PR?

If anything else needs to be addressed for this PR, please let me know. Thanks.


<!-- Sent from Reviewable.io -->

devbww reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131)* status: 0 of 12 files reviewed, 9 unresolved discussions (waiting on @jsrinnn)

___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc` line 48 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR5-r3-48:-NSRqQcCAgZ1wJ8NnBCh:b-u7rzi3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc#L48)):*
<details><summary><i>Previously, jsrinnn wrote…</i></summary><blockquote>

Ack. 

Can you give approval for the changes in this PR?

If anything else needs to be addressed for this PR, please let me know. Thanks.
</blockquote></details>

[Sorry our last messages crossed.]

Yes.  I've only just glanced at the moment, but it's looking good.

If you're not using `SubMessage()` yet then the convention to pass `name` (and `indent`) to `T::DebugString()` may not affect you.  But if you don't yet have #11197 you'll probably want to rebase again to make sure.


<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131)* status: 0 of 12 files reviewed, 9 unresolved discussions (waiting on @devbww)

___
*[`google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc` line 48 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NRd__cgm2ZDsjn3khR5-r3-48:-NSRrJbPA-vXc-uUZjUD:b24ctlv) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08f75b30a9d505381344c9f802782321773a1a19/google/cloud/bigquery/v2/minimal/internal/bigquery_http_response.cc#L48)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

[Sorry our last messages crossed.]

Yes.  I've only just glanced at the moment, but it's looking good.

If you're not using `SubMessage()` yet then the convention to pass `name` (and `indent`) to `T::DebugString()` may not affect you.  But if you don't yet have #11197 you'll probably want to rebase again to make sure.

</blockquote></details>

Thanks. Right now I am not using SubMessage()

Yeah I just noticed compile errors after the rebase.

I'll go ahead and fix them.

Will update the review thread once I have pushed the changes.


<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
I have fixed all the compile errors and made changes to response logging corresponding to the DebugString signature change in google::cloud library.

Bradley, can you review the latest changes and approve when you get a chance? I believe all feedback has been addressed. If there are any additional changes for this PR, let me know.

Thanks!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NSS9lQO3CSVjpfd7VoR:b3p3r8h)* status: 0 of 12 files reviewed, 9 unresolved discussions (waiting on @devbww)



<!-- Sent from Reviewable.io -->

devbww reviewed (approved):
> Reviewed 3 of 12 files at r2, 9 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131)* status: all files reviewed, 5 unresolved discussions (waiting on @jsrinnn)

___
*[`google/cloud/bigquery/v2/minimal/internal/job.cc` line 43 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdShT0Az3RUvvAlOmV:-NSdShT1-8njW1MSvSqD:b-27l4n4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job.cc#L43)):*
> ```cpp
>       .StringField("location", std::move(location))
>       .Build();
> }
> ```

It looks like these should be `DebugString()` functions on `JobConfiguration` and `JobReference`, but that can wait for a followup.
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 91 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdYhtvEiANQWxbtkaF:-NSdYhtvEiANQWxbtkaG:b-vw6g44) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L91)):*
> ```cpp
>   return internal::DebugFormatter(name, options, indent)
>       .StringField("http_response",
>                    http_response.DebugString("BigQueryHttpResponse", options))
> ```

Missing passing `indent`.
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 92 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdVRjcDpWx92N6CT-d:-NSdVRjcDpWx92N6CT-e:b7zcw9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L92)):*
> ```cpp
>       .StringField("http_response",
>                    http_response.DebugString("BigQueryHttpResponse", options))
>       .StringField("job", job.DebugString("Job", options, indent))
> ```

It looks like `http_response` and `job` could use `SubMessage()`.

This would also correct their output format to, for example, "http_response { ... }" instead of "http_response: \"BigQueryHttpResponse { ... }\"".
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 131 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdVkuw1FcyvZl_XY6R:-NSdVkuw1FcyvZl_XY6S:baj8hyx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L131)):*
> ```cpp
>   return internal::DebugFormatter(name, options, indent)
>       .StringField("http_response",
>                    http_response.DebugString("BigQueryHttpResponse", options))
> ```

It looks like `http_response` could use `SubMessage()`.
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response_test.cc` line 240 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdZkTf4UIhJvHAdKQF:-NSdZkTf4UIhJvHAdKQG:ba4rwrv) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response_test.cc#L240)):*
> ```cpp
>             R"( query: "select 1;")"
>             R"( }")"
>             R"( job_r...<truncated>...")"
> ```

This demonstrates the problems with passing a submessage as a string field: it gets truncated  as a string field when it isn't one, and the braces aren't matched.


<!-- Sent from Reviewable.io -->

jsrinnn reviewed (commented):
Addressed all feeback. Thank you Bradley!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-:-NSfiaQWAqLgaxA6y88H:b8hxh21)* status: 7 of 13 files reviewed, all discussions resolved (waiting on @devbww)

___
*[`google/cloud/bigquery/v2/minimal/internal/job.cc` line 43 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdShT0Az3RUvvAlOmV:-NSfSSPS2cSeX2e9KsD6:b-hiftmp) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job.cc#L43)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like these should be `DebugString()` functions on `JobConfiguration` and `JobReference`, but that can wait for a followup.
</blockquote></details>

Done

I addressed it in this PR. I have added DebugString for JobConfiguration and JobReference so we can use the SubMessage()
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 91 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdYhtvEiANQWxbtkaF:-NSfUBEC4hoPUGyWXC8i:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L91)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Missing passing `indent`.
</blockquote></details>

Done
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 92 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdVRjcDpWx92N6CT-d:-NSfVcxV2n14f036jhzv:b-hqgm3b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L92)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like `http_response` and `job` could use `SubMessage()`.

This would also correct their output format to, for example, "http_response { ... }" instead of "http_response: \"BigQueryHttpResponse { ... }\"".
</blockquote></details>

Done

Ah...nice. Thanks!
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response.cc` line 131 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdVkuw1FcyvZl_XY6R:-NSfVgRO3VUzxKv1fvmm:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response.cc#L131)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like `http_response` could use `SubMessage()`.
</blockquote></details>

Done
___
*[`google/cloud/bigquery/v2/minimal/internal/job_response_test.cc` line 240 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11131#-NSdZkTf4UIhJvHAdKQF:-NSfWC1W0hMi7C7xtIxA:b-thu12y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a66b5b8136baca80c8eeff853ff90f3b466d880f/google/cloud/bigquery/v2/minimal/internal/job_response_test.cc#L240)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This demonstrates the problems with passing a submessage as a string field: it gets truncated  as a string field when it isn't one, and the braces aren't matched.
</blockquote></details>

Done

This should now be addressed due to incorporating SubMessage as per previous feedback.


<!-- Sent from Reviewable.io -->
